docs/design/gh-ost-integration.md

```diff
@@ -0,0 +1,507 @@
+
+However, we need to refactor pre-migration validation and migration status, because the current logic doesn't allow migration to have the same version, rerun will lead to failure. We can refactor to reuse version to keep it idempotent.
+
+Another approach is to abandon the previous version and start a new issue with a new version. The benefit is that we don't have to support rerun, so it's less complex.
+
+For the first iteration, we just let it fail.
+
+### After Migration
+
+After migration is finished, we will have `tablename` and `_tablename_del`, the latter is the original table. We are too scared to delete the original table for the user, so we'll generate a task to delete the table if approved by user.
+
+### Task dependency
+
+Tasks under the same stage run one by one by now. With gh-ost integration, we introduce task dependency. A task can only be running if all of its dependencies are done. Tasks that are not blocked by dependencies can run in parallel.
+
+To store the dependency relationship, we can add a field in Task.
+
+```Go
+type Task struct {
+    // BlockedBy stores TaskIDs
+    // Task cannot be running if not all of the BlockedBy tasks are done
+    BlockedBy []int
+}
+```
+
+However, we can only know the `TaskID` of a `Task` after insertion, so we have to update `Task` after all tasks in the same stage are inserted into the database.
+
+### MigrationContext
+
+We will introduce a new issue type for gh-ost. This is how we tell the difference between gh-ost and direct DDL.
+
+And a new task payload type and task executor for each execution task.
+
+```Go
+// issue related
+
+const IssueDatabaseSchemaUpdateGhost IssueType = "bb.issue.database.schema.update.ghost"
+
+type UpdateSchemaGhostDetail struct {
+    DatabaseID int `json:"databaseId"`
+    DatabaseName string `json:"databaseName"`
+    Statement string `json:"statement"`
+    EarliestAllowedTs int64 `jsonapi:"attr,earliestAllowedTs"`
+}
+
+type UpdateSchemaGhostContext struct {
+    MigrationType db.MigrationType `json:"migrationType"`
+    UpdateSchemaDetailList []*UpdateSchemaGhostDetail `json:"updateSchemaDetailList`
+    VCSPushEvent *vcs.PushEvent
+}
+
+// task related
+
+const TaskDatabaseSchemaUpdateGhost TaskType = "bb.task.database.schema.update.ghost"
+type TaskDatabaseSchemaUpdateGhostPayload struct {
+    MigrationType db.MigrationType `json:"migrationType,omitempty"`
+    Statement     string           `json:"statement,omitempty"`
+    SchemaVersion string           `json:"schemaVersion,omitempty"`
+    VCSPushEvent  *vcs.PushEvent   `json:"pushEvent,omitempty"`
+    // more to come
+}
+
+const TaskDatabaseSchemaUpdateGhostCutover = "bb.task.database.schema.update.ghost.cutover"
+type TaskDatabaseSchemaUpdateGhostCutoverPayload struct {
+    // ParentTaskID references back to the gh-ost executor task ID
+    ParentTaskID int `json:"parentTaskId,omitempty"`
```

Comment:
I am thinking if we can get rid of the cutover task at all and thus avoid introducing the parent and dependency stuff.

The main ghost task will do the sync in the background. And we can have a task check to check whether the table has been synced and ready to do cutover. Only the task check passes, the user can click the approval button.

The approval might still fail because of the underlying ghost sanity check when actually performing the cutover. But that's ok and we can retry the task again later.

@RainbowDashy @d-bytebase thoughts?

Reply:
Resolved offline